crates/accelerate/src/sparse_pauli_op.rs

```diff
@@ -337,163 +338,463 @@ pub fn decompose_dense(
     };
-    let mid = 1usize << (num_qubits - 1);
-    recurse_if_nonzero(
-        Pauli::I,
-        0.5 * factor,
-        &block.slice(s![..mid, ..mid]) + &block.slice(s![mid.., mid..]),
-    );
-    recurse_if_nonzero(
-        Pauli::X,
-        0.5 * factor,
-        &block.slice(s![..mid, mid..]) + &block.slice(s![mid.., ..mid]),
-    );
-    recurse_if_nonzero(
-        Pauli::Y,
-        0.5 * Complex64::i() * factor,
-        &block.slice(s![..mid, mid..]) - &block.slice(s![mid.., ..mid]),
-    );
-    recurse_if_nonzero(
-        Pauli::Z,
-        0.5 * factor,
-        &block.slice(s![..mid, ..mid]) - &block.slice(s![mid.., mid..]),
-    );
+    out.phases.push(phase);
+    out.coeffs.push(value);
+}
+
+/// Pack the information about which row, column and qubit we're considering into a single `usize`.
+/// Complex64 data is 16 bytes long and the operators are square and must be addressable in memory,
+/// so the row and column are hardware limited to be of width `usize::BITS / 2 - 2` each.  However,
+/// we don't need to store at a granularity of 1, because the last 2x2 block we handle manually, so
+/// we can remove an extra least significant bit from the row and column.  Regardless of the width
+/// of `usize`, we can therefore track the state for up to 30 qubits losslessly, which is greater
+/// than the maximum addressable memory on a 64-bit system.
+///
+/// For a 64-bit usize, the bit pattern is stored like this:
+///
+///    0b__000101__11111111111111111111111110000__11111111111111111111111110000
+///        <-6-->  <------------29------------->  <------------29------------->
+///          |                  |                              |
+///          |         uint of the input row         uint of the input column
+///          |         (once a 0 is appended)        (once a 0 is appended)
+///          |
+///        current qubit under consideration
+///
+/// The `qubit` field encodes the depth in the call stack that the user of the `PauliLocation`
+/// should consider.  When the stack is initialised (before any calculation is done), it starts at
+/// the highest qubit index (`num_qubits - 1`) and decreases from there until 0.
+///
+/// The `row` and `col` methods form the top-left corner of a `(2**(qubit + 1), 2**(qubit + 1))`
+/// submatrix (where the top row and leftmost column are 0).  The least significant `qubit + 1`
+/// bits of the of row and column are therefore always zero; the 0-indexed qubit still corresponds
+/// to a 2x2 block.  This is why we needn't store it.
+#[derive(Debug, Clone, Copy)]
+struct PauliLocation(usize);
+
+impl PauliLocation {
+    const QUBIT_SHIFT: u32 = usize::BITS - 6;
+    const QUBIT_MASK: usize = (usize::MAX >> Self::QUBIT_SHIFT) << Self::QUBIT_SHIFT;
+    const ROW_SHIFT: u32 = usize::BITS / 2 - 3;
+    const ROW_MASK: usize =
+        ((usize::MAX >> Self::ROW_SHIFT) << Self::ROW_SHIFT) & !Self::QUBIT_MASK;
+    const COL_SHIFT: u32 = 0;
```

Comment:
why is this const needed?

Reply:
It's just for consistency with the `row` ones, so the functions all look the same.

Reply:
Done in e2a97e4.